README.md
## Goodnews

## Disclaimer: Proper Attribution of News Sources for Scraped Content

*This news scraping application is designed to aggregate and present news articles from various sources across the internet. While we strive to provide accurate and up-to-date information, it is important to note that the responsibility for proper attribution of news sources lies with each individual user of the application.*

*By using this application, you acknowledge that it is your responsibility to always mention and attribute the original source of any news content that is obtained through the scraping process. This includes providing appropriate credit to the authors, publications, and websites from which the news articles originate.*

*We encourage all users to respect the intellectual property rights of news publishers and content creators. Failure to attribute sources correctly may infringe upon copyright and ethical guidelines. Additionally, inaccurate or misleading attribution can undermine the credibility and integrity of the news dissemination process.*

This GO package is scrapping websites with good news, saving the news from the last 2 days (in case they were not already saved) to the SQLite3 DB (by default `./data/news_items.db`) and sending them to the external service, specified in the `external.SendToExternalService` function (Telegram by default), sorted by the posted date. Once the item is sent, it is marked accordingly in the db and won't be sent again.

### Pre-req and installation

To run/build the project, you need GO 1.20+ and the packages mentioned in `go.mod`. In all the below commands, do not forget to replace `<YOUR_TELEGRAM_BOT_API_KEY>` and `<ID_OF_THE_CHAT_TO_SEND_NEWS_TO>` placeholders with the actual values.

#### To run the project:

`mkdir data && API_KEY=<YOUR_TELEGRAM_BOT_API_KEY> CHAT_ID=<ID_OF_THE_CHAT_TO_SEND_NEWS_TO> go run main.go`

#### To build and run the project:

```
go build -o goodnews
chmod +x ./goodnews
export API_KEY=<YOUR_TELEGRAM_BOT_API_KEY>
export CHAT_ID=<ID_OF_THE_CHAT_TO_SEND_NEWS_TO>
mkdir data
./goodnews
```

#### To build and run with Docker:

```
docker build -t goodnews:<VERSION> .
docker volume create goodnews_data
docker run --rm -v goodnews_data:/app/data -e API_KEY=<YOUR_TELEGRAM_BOT_API_KEY> -e CHAT_ID=<ID_OF_THE_CHAT_TO_SEND_NEWS_TO> goodnews:<VERSION>
```

If you already have a SQLite3 db file, and you'd like to use it with the newly created docker container, you can run the following command from the directory with the file to copy the file to the volume before you run the container (make sure that you have jq installed in your system):

`cp ./news_items.db $(docker volume inspect goodnews_data | jq -r '.[0].Mountpoint')/news_items.db`

#### Additional information

Since version 1.1.0 you can use the following flags (both of them are turned off by default):

* `--dry-run` to perform a dry run. We are still going to scrape the sources, but no write actions will be done to DB and message won't be sent to external source
* `--debug` to output more information during the run

Usage example:

`go run main.go --dry-run --debug`


CHANGELOG.md
# 2.1.0

* Changed formatting to html instead of markdown to support more options
* Moved source to the title
* Adjusted `external` package to handle 2 edge-cases: remove asterisk to avoid 400 response; separate P1 in case it is too long

# 2.0.2

* Fixed a bug when text[i] items of the news were not considered for a caption

# 2.0.1

* Fixed caption indentation
* Removed datetime in caption by default 
* Added emojis before the link

# 2.0.0

* Breaking change! Introduced `scraping.ScrapeEntity` struct which is now a part of the `NewScraper` constructor. This change allows to define additional news sources without adjusting `scraping` funcs
* Added `scraping.PickRandomUserAgent` func to randomize user agent for each run
* Introduced `database.CheckIfRecordWithUrlExists` func to avoid scraping news that already exist in DB
* Added a new source
* Reduced the value of newsAgeDays to 2 days instead of 14

# 1.2.0

* Improved the project structure for better readability: separated structs and functions
* Introduced `external.assembleCaption` message formatting func to respect Telegram API limitations
* Adjusted signatures of `database.CheckAndInsertItem` and `database.ProcessUnsentItems` to include additional args
* Added more items to `external.pickRandomMessageEnding`

# 1.1.1

* Introduced 0.5sec delay between colly requests to avoid potential rate-limiting

# 1.1.0

* Added README and CHANGELOG
* Refactored `GetNewsUrls`, `GetNews` functions to support multiple sources
* Refactored `formatTime` to support dynamic layouts
* Added `https://ntdtv.ru/c/pozitivnye-novosti` as a secondary news source
* Added `pickRandomMessageEnding` function to randomize message endings
* Added dry-run and debug modes available with the flags `--dry-run` and `--debug`

# 1.0.0

* Initial version


main.go
package main

import (
	"flag"
	"log"

	"goodnews/database"
	"goodnews/scraping"
)

var dryRun bool
var debug bool
var newsItems []scraping.NewsItem

func main() {
	flag.BoolVar(&dryRun, "dry-run", false, "Perform a dry run. We are still going to scrape the sources, but no write actions will be done to DB and message won't be sent to external source")
	flag.BoolVar(&debug, "debug", false, "Output more information during the run")
	flag.Parse()
	db, err := database.InitDB(dryRun, "data/news_items.db")

	if err != nil {
		log.Fatal(err)
	}
	defer db.Close()

	scrapeEntities := []scraping.ScrapeEntity{
		{
			SourceUrl: "https://positivnews.ru/",
			ScrapeNewsUrlsElements: scraping.ScrapeNewsURL{
				UrlElements: []string{"div.digital-newspaper-container", "article.post"},
			},
			ScrapeNewsHTMLElements: scraping.ScrapeNewsHTML{
				TextTxt:      ".entry-content p",
				CategoryTxt:  ".post-categories a",
				PostedAttr:   []string{".entry-meta time.updated", "datetime"},
				PostedFormat: "2006-01-02T15:04:05-07:00",
				TitleTxt:     ".entry-title",
				ImageAttr:    []string{"div.post-inner div.post-thumbnail img.wp-post-image", "src"},
			},
		},
		{
			SourceUrl: "https://ntdtv.ru/c/pozitivnye-novosti",
			ScrapeNewsUrlsElements: scraping.ScrapeNewsURL{
				UrlElements: []string{"div.entry-image"},
			},
			ScrapeNewsHTMLElements: scraping.ScrapeNewsHTML{
				TextTxt:      "div[id=cont_post] p",
				CategoryTxt:  "div.entry-meta a[href=\"https://ntdtv.ru/\"]",
				PostedAttr:   []string{"span.entry-date time", "datetime"},
				PostedFormat: "2006-01-02 15:04:05",
				TitleTxt:     "header.entry-header h1",
				ImageAttr:    []string{"link[itemprop=thumbnailUrl]", "href"},
			},
		},
		{
			SourceUrl: "https://allpozitive.ru/",
			ScrapeNewsUrlsElements: scraping.ScrapeNewsURL{
				UrlElements: []string{"div.col-1-2.mq-sidebar div.sb-widget ul.cp-widget.row.clearfix li.cp-wrap.clearfix div.cp-data p.cp-widget-title"},
			},
			ScrapeNewsHTMLElements: scraping.ScrapeNewsHTML{
				TextTxt:      "div.entry.clearfix",
				CategoryTxt:  "header.post-header p.meta.post-meta a[rel=\"category tag\"]",
				PostedAttr:   []string{"p.meta.post-meta", "datetime"},
				PostedFormat: "2006-01-02T15:04:05-07:00",
				TitleTxt:     "h1.post-title",
				ImageAttr:    []string{"div.post-thumbnail img", "src"},
				PostedTextToParse: scraping.TextToParse{
					Regex:  `\d{2}\.\d{2}\.\d{4}`,
					Layout: "02.01.2006",
				},
			},
		},
	}

	s := scraping.NewScraper(scraping.PickRandomUserAgent(), scrapeEntities, 500, debug)
	newsUrls := s.ScrapeNewsUrlsFromSources()
	var newsUrlsNotAlreadyInDB []string

	for i := 0; i < len(newsUrls); i++ {
		urlExists, err := database.CheckIfRecordWithUrlExists(dryRun, debug, db, newsUrls[i])
		if err != nil {
			log.Printf("Error checking the url %s in the db: %v\n", newsUrls[i], err)
		}

		if !urlExists {
			newsUrlsNotAlreadyInDB = append(newsUrlsNotAlreadyInDB, newsUrls[i])
		}
	}
	if debug {
		log.Printf("DEBUG: len(newsUrlsNotAlreadyInDB): %d", len(newsUrlsNotAlreadyInDB))
		log.Printf("DEBUG: newsUrlsNotAlreadyInDB: %v", newsUrlsNotAlreadyInDB)
	}

	newsItems, err = s.ScrapeNewsFromNewsUrls(newsUrlsNotAlreadyInDB)

	if err != nil {
		log.Printf("Error from s.ScrapeNewsFromNewsUrls: %v\n", err)
	}

	if debug {
		log.Printf("DEBUG: Total number of news received: %d. Processing the news...", len(newsUrls))
		for i := 0; i < len(newsUrls); i++ {
			log.Printf("DEBUG: Item # %d: %s", i, newsUrls[i])
		}
	}

	for _, item := range newsItems {
		if debug {
			log.Printf("DEBUG: %s, %s, %s, %s, %s, p1:%s\n\ntext(elements: %d):%v\n\ntext[0]:%s", item.Url, item.Category, item.Posted, item.Title, item.Image, item.P1, len(item.Text), item.Text, item.Text[0])
		}
		err := database.CheckAndInsertItem(dryRun, db, item, 2)
		if err != nil {
			log.Printf("Error processing item: %v", err)
		}
	}

	log.Println("Running processUnsentItems...")

	err = database.ProcessUnsentItems(dryRun, db, 500)
	if err != nil {
		log.Printf("Error processing unsent items: %v", err)
	}
}


database/database.go
package database

import (
	"database/sql"
	"encoding/json"
	"goodnews/external"
	"goodnews/scraping"
	"log"
	"sort"
	"time"

	_ "github.com/mattn/go-sqlite3"
)

func InitDB(dryRun bool, filePath string) (*sql.DB, error) {
	db, err := sql.Open("sqlite3", filePath)

	if !dryRun {
		createTableSQL := `
            CREATE TABLE IF NOT EXISTS news_items (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                url TEXT,
                category TEXT,
                posted TEXT,
                title TEXT,
                image TEXT,
                text TEXT,
                p1 TEXT,
                item_was_sent BOOLEAN
            );
        `

		_, err = db.Exec(createTableSQL)
		if err != nil {
			log.Fatal(err)
		}
		log.Println("Database setup completed")
	}

	if err != nil {
		return nil, err
	}

	return db, nil
}

func CheckAndInsertItem(dryRun bool, db *sql.DB, item scraping.NewsItem, newsAgeDays int) error {
	if !dryRun {
		query := "SELECT COUNT(*) FROM news_items WHERE url = ?"
		var count int
		err := db.QueryRow(query, item.Url).Scan(&count)
		if err != nil {
			return err
		}

		currentTime := time.Now()

		postedTime, err := time.Parse("02-01-2006 15:04:05", item.Posted)
		if err != nil {
			return err
		}

		daysDiff := currentTime.Sub(postedTime).Hours() / 24

		if count > 0 {
			log.Printf("Item with URL '%s' already exists in the database.", item.Url)
			return nil
		}

		if count == 0 && daysDiff <= float64(newsAgeDays) {
			textJSON, err := json.Marshal(item.Text)
			if err != nil {
				return err
			}

			insertQuery := `
		INSERT INTO news_items (url, category, posted, title, image, text, p1, item_was_sent)
		VALUES (?, ?, ?, ?, ?, ?, ?, ?)
	`
			_, err = db.Exec(insertQuery, item.Url, item.Category, item.Posted, item.Title, item.Image, textJSON, item.Text[0], false)
			if err != nil {
				return err
			}

			log.Printf("Item with URL '%s' inserted into the database.", item.Url)
		}
	} else {
		log.Printf("DRY-RUN: Item with URL '%s' inserted into the database.", item.Url)
	}
	return nil
}

func ProcessUnsentItems(dryRun bool, db *sql.DB, reqSleepMs int) error {
	if !dryRun {
		tx, err := db.Begin()

		if err != nil {
			return err
		}
		defer tx.Rollback()

		query := "SELECT id, category, posted, url, title, image, text, p1 FROM news_items WHERE item_was_sent = false"
		rows, err := tx.Query(query)
		if err != nil {
			return err
		}
		defer rows.Close()

		var unsentItems []scraping.NewsItem

		for rows.Next() {
			var id int
			var category, posted, url, title, image, p1 string
			var textJSON []byte
			var text []string

			if err := rows.Scan(&id, &category, &posted, &url, &title, &image, &textJSON, &p1); err != nil {
				return err
			}

			if err := json.Unmarshal(textJSON, &text); err != nil {
				return err
			}

			item := scraping.NewsItem{
				Id:       id,
				Category: category,
				Posted:   posted,
				Url:      url,
				Title:    title,
				Image:    image,
				Text:     text,
				P1:       p1,
			}

			unsentItems = append(unsentItems, item)
		}

		if len(unsentItems) == 0 {
			log.Println("No unsent items found...")
			return nil
		}

		sort.Slice(unsentItems, func(i, j int) bool {
			timeI, _ := time.Parse("02-01-2006 15:04:05", unsentItems[i].Posted)
			timeJ, _ := time.Parse("02-01-2006 15:04:05", unsentItems[j].Posted)
			return timeI.Before(timeJ)
		})

		for _, item := range unsentItems {
			err := external.SendToExternalService(item)
			time.Sleep(time.Duration(reqSleepMs) * time.Millisecond)
			if err != nil {
				log.Printf("Error sending item with ID %d to external service: %v", item.Id, err)
				continue
			}

			updateQuery := "UPDATE news_items SET item_was_sent = true WHERE id = ?"
			_, err = tx.Exec(updateQuery, item.Id)
			if err != nil {
				log.Printf("Error updating item_was_sent flag for item with ID %d: %v", item.Id, err)
				tx.Rollback()
				continue
			}
		}

		if err := tx.Commit(); err != nil {
			return err
		}
	} else {
		log.Printf("DRY-RUN: Finished processUnsentItems")
	}

	return nil
}

func CheckIfRecordWithUrlExists(dryRun bool, debug bool, db *sql.DB, url string) (bool, error) {
	if !dryRun {
		query := "SELECT COUNT(*) FROM news_items WHERE url = ?"
		var count int
		err := db.QueryRow(query, url).Scan(&count)
		if debug {
			log.Printf("DEBUG: looked in db for item with %s url found %d\n", url, count)
		}
		if err != nil {
			return false, err
		}

		if count > 0 {
			return true, nil
		} else {
			return false, nil
		}
	} else {
		return true, nil
	}
}


scraping/scraping.go
package scraping

import (
	"fmt"
	"log"
	"math/rand"
	"net/url"
	"regexp"
	"strings"
	"time"

	"github.com/gocolly/colly"
)

type NewsItem struct {
	Id                                      int
	Url, Category, Posted, Title, Image, P1 string
	Text                                    []string
}

type ScrapeEntity struct {
	SourceUrl              string
	ScrapeNewsUrlsElements ScrapeNewsURL
	ScrapeNewsHTMLElements ScrapeNewsHTML
}

type ScrapeNewsURL struct {
	UrlElements []string
}

type TextToParse struct {
	Regex, Layout string
}

type ScrapeNewsHTML struct {
	TextTxt, CategoryTxt, PostedFormat, TitleTxt string
	PostedAttr, ImageAttr                        []string
	PostedTextToParse                            TextToParse
}

type Scraper struct {
	UserAgent      string
	Collector      *colly.Collector
	ReqSleepMs     int
	DebugFlag      bool
	ScrapeEntities []ScrapeEntity
}

func NewScraper(userAgent string, ScrapeEntities []ScrapeEntity, reqSleepMs int, debug bool) *Scraper {
	c := colly.NewCollector()
	c.UserAgent = userAgent

	return &Scraper{
		UserAgent:      userAgent,
		Collector:      c,
		ReqSleepMs:     reqSleepMs,
		DebugFlag:      debug,
		ScrapeEntities: ScrapeEntities,
	}
}

func (s *Scraper) ScrapeNewsUrlsFromSources() []string {
	var newsUrls []string

	for i := 0; i < len(s.ScrapeEntities); i++ {

		s.Collector.OnError(func(_ *colly.Response, err error) {
			log.Println("Something went wrong: ", err)
		})

		for _, htmlElement := range s.ScrapeEntities[i].ScrapeNewsUrlsElements.UrlElements {
			s.Collector.OnHTML(htmlElement, func(e *colly.HTMLElement) {
				newsUrls = append(newsUrls, e.ChildAttr("a", "href"))
			})
			s.Collector.Visit(s.ScrapeEntities[i].SourceUrl)
			time.Sleep(time.Duration(s.ReqSleepMs) * time.Millisecond)
		}
	}

	newsUrls = s.removeDuplicatesAndRootSite(newsUrls)

	if s.DebugFlag {
		log.Printf("DEBUG: newsUrls: %v", newsUrls)
	}

	return newsUrls
}

func (s *Scraper) ScrapeNewsFromNewsUrls(newsUrls []string) ([]NewsItem, error) {
	if len(newsUrls) == 0 {
		return nil, fmt.Errorf("the NewsUrls is empty. Please make sure to run scraper.ScrapeNewsUrlsFromSources() first")
	} else {
		var newsItems []NewsItem

		for i := 0; i < len(newsUrls); i++ {
			var newsItem NewsItem
			newsItem.Url = newsUrls[i]

			for j := 0; j < len(s.ScrapeEntities); j++ {
				u, err := url.Parse(s.ScrapeEntities[j].SourceUrl)
				if err != nil {
					log.Printf("Error! the url %s can't be parsed! Proceeding without it", s.ScrapeEntities[j].SourceUrl)
					break
				}

				if strings.Contains(newsUrls[i], u.Scheme+"://"+u.Host) {
					s.Collector.OnError(func(_ *colly.Response, err error) {
						log.Println("Something went wrong: ", err)
					})

					s.Collector.OnHTML(s.ScrapeEntities[j].ScrapeNewsHTMLElements.TextTxt, func(e *colly.HTMLElement) {
						// Workaround for the articles without the heading <p> element in the div with the post
						text := strings.TrimSpace(e.Text)
						if text != "" {
							newsItem.Text = append(newsItem.Text, text)
						}
					})

					s.Collector.OnHTML(s.ScrapeEntities[j].ScrapeNewsHTMLElements.CategoryTxt, func(e *colly.HTMLElement) {
						newsItem.Category = e.Text
					})

					s.Collector.OnHTML(s.ScrapeEntities[j].ScrapeNewsHTMLElements.TitleTxt, func(e *colly.HTMLElement) {
						newsItem.Title = e.Text
					})

					s.Collector.OnHTML(s.ScrapeEntities[j].ScrapeNewsHTMLElements.PostedAttr[0], func(e *colly.HTMLElement) {
						newsItem.Posted = formatTime(e.Attr(s.ScrapeEntities[j].ScrapeNewsHTMLElements.PostedAttr[1]), s.ScrapeEntities[j].ScrapeNewsHTMLElements.PostedFormat, true)
						if len(newsItem.Posted) == 0 && len(e.Text) > 0 {
							parsedDate, err := parseDateTimeFromTextFallback(e.Text, s.ScrapeEntities[j].ScrapeNewsHTMLElements.PostedTextToParse.Regex, s.ScrapeEntities[j].ScrapeNewsHTMLElements.PostedTextToParse.Layout, s.ScrapeEntities[j].ScrapeNewsHTMLElements.PostedFormat)
							if err != nil {
								fmt.Println("Error:", err)
							} else {

								newsItem.Posted = parsedDate
							}
						}
					})

					s.Collector.OnHTML(s.ScrapeEntities[j].ScrapeNewsHTMLElements.ImageAttr[0], func(e *colly.HTMLElement) {
						newsItem.Image = e.Attr(s.ScrapeEntities[j].ScrapeNewsHTMLElements.ImageAttr[1])
					})
					s.Collector.Visit(newsUrls[i])

					if s.DebugFlag {
						log.Printf("DEBUG: i = %d, processing url: %s len(newsUrls): %d\n", i, newsUrls[i], len(newsUrls))
					}

					time.Sleep(time.Duration(s.ReqSleepMs) * time.Millisecond)

					if len(newsItem.Text) > 1 {
						newsItem.P1 = newsItem.Text[0] + " " + newsItem.Text[1]
					} else {
						newsItem.P1 = newsItem.Text[0]
					}
					newsItems = append(newsItems, newsItem)
					break
				}
			}
		}
		return newsItems, nil
	}
}

func formatTime(src, inputLayout string, suppressError bool) string {

	t, err := time.Parse(inputLayout, src)
	if err != nil {
		if !suppressError {
			fmt.Println("Error parsing timestamp:", err)
		}
		return ""
	}

	return t.Format("02-01-2006 15:04:05")
}

func (s *Scraper) removeDuplicatesAndRootSite(newsUrls []string) []string {
	uniqueMap := make(map[string]bool)
	result := []string{}

	for i := 0; i < len(s.ScrapeEntities); i++ {
		uniqueMap[s.ScrapeEntities[i].SourceUrl] = true
	}

	for _, url := range newsUrls {
		if url != "" && !uniqueMap[url] {
			uniqueMap[url] = true
			result = append(result, url)
		}
	}
	return result
}

func PickRandomUserAgent() string {
	userAgents := []string{
		"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36",
		"Mozilla/5.0 (Macintosh; Intel Mac OS X 13_5_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36",
		"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36",
		"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36 Edg/116.0.1938.76",
		"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36 Edg/116.0.1938.62",
		"Mozilla/5.0 (Macintosh; Intel Mac OS X 13_5_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36 Edg/116.0.1938.76",
		"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36 OPR/102.0.0.0",
		"Mozilla/5.0 (Macintosh; Intel Mac OS X 13_5_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36 OPR/102.0.0.0",
		"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36 OPR/102.0.0.0",
		"Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:109.0) Gecko/20100101 Firefox/117.0",
		"Mozilla/5.0 (Macintosh; Intel Mac OS X 13.5; rv:109.0) Gecko/20100101 Firefox/117.0",
		"Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/117.0",
	}

	rand.New(rand.NewSource(time.Now().UnixNano()))
	randomIndex := rand.Intn(len(userAgents))

	return userAgents[randomIndex]
}

func parseDateTimeFromTextFallback(inputText, regex, layout, timeFormat string) (string, error) {
	re := regexp.MustCompile(regex)
	matches := re.FindStringSubmatch(inputText)

	if len(matches) == 0 {
		return "", fmt.Errorf("no date found in the input string")
	}

	parsedDate, err := time.Parse(layout, matches[0])
	if err != nil {
		return "", err
	}

	formattedDateTime := parsedDate.Format(timeFormat)
	return formatTime(formattedDateTime, timeFormat, false), nil
}


goodnews.go
package main

import (
	"database/sql"
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"log"
	"math/rand"
	"net/http"
	"net/url"
	"os"
	"sort"
	"strings"
	"time"

	"github.com/gocolly/colly"
	_ "github.com/mattn/go-sqlite3"
)

type NewsItem struct {
	id                                      int
	url, category, posted, title, image, p1 string
	text                                    []string
}

var dryRun bool
var debug bool

func main() {
	flag.BoolVar(&dryRun, "dry-run", false, "Perform a dry run. We are still going to scrape the sources, but no write actions will be done to DB and message won't be sent to external source")
	flag.BoolVar(&debug, "debug", false, "Output more information during the run")
	flag.Parse()
	var news []NewsItem
	urls := []string{"https://positivnews.ru/", "https://ntdtv.ru/c/pozitivnye-novosti"}
	userAgent := "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36 Edg/116.0.1938.62"
	db, err := sql.Open("sqlite3", "data/news_items.db")

	if !dryRun {
		createTableSQL := `
		CREATE TABLE IF NOT EXISTS news_items (
			id INTEGER PRIMARY KEY AUTOINCREMENT,
			url TEXT,
			category TEXT,
			posted TEXT,
			title TEXT,
			image TEXT,
			text TEXT,
			p1 TEXT,
			item_was_sent BOOLEAN
		);
	`

		_, err = db.Exec(createTableSQL)
		if err != nil {
			log.Fatal(err)
		}
		log.Println("Database setup completed")
	}

	if err != nil {
		log.Fatal(err)
	}
	defer db.Close()

	var newsUrl []string

	for _, url := range urls {
		newsUrl = append(newsUrl, removeDuplicatesAndRootSite(GetNewsUrls(url, userAgent), url)...)
	}

	log.Printf("Total number of news received: %d. Processing the news...", len(newsUrl))

	for i := 0; i < len(newsUrl); i++ {
		if debug {
			log.Printf("DEBUG: Item # %d: %s", i, newsUrl[i])
		}
		news = append(news, GetNews(newsUrl[i], userAgent))
	}

	for _, item := range news {
		if debug {
			log.Printf("DEBUG: %s, %s, %s, %s, %s, p1:%s\n\ntext(elements: %d):%v\n\ntext[0]:%s", item.url, item.category, item.posted, item.title, item.image, item.p1, len(item.text), item.text, item.text[0])
		}
		err := checkAndInsertItem(db, item)
		if err != nil {
			log.Printf("Error processing item: %v", err)
		}
	}

	fmt.Println("Running processUnsentItems...")

	err = processUnsentItems(db)
	if err != nil {
		log.Printf("Error processing unsent items: %v", err)
	}
}

func GetNewsUrls(url, ua string) []string {
	c := colly.NewCollector()
	c.UserAgent = ua
	urls := []string{}

	c.OnError(func(_ *colly.Response, err error) {
		log.Println("Something went wrong: ", err)
	})

	switch url {
	case "https://positivnews.ru/":
		c.OnHTML("div.digital-newspaper-container", func(e *colly.HTMLElement) {
			urls = append(urls, e.ChildAttr("a", "href"))
		})

		c.OnHTML("article.post", func(e *colly.HTMLElement) {
			urls = append(urls, e.ChildAttr("a", "href"))
		})
	case "https://ntdtv.ru/c/pozitivnye-novosti":
		c.OnHTML("div.entry-image", func(e *colly.HTMLElement) {
			urls = append(urls, e.ChildAttr("a", "href"))
		})
	default:
		log.Panic("The news source passed to GetNewsUrls func is not found!")
	}

	c.Visit(url)

	return urls
}

func GetNews(url, ua string) NewsItem {
	news := NewsItem{}
	news.url = url
	c := colly.NewCollector()
	c.UserAgent = ua

	c.OnError(func(_ *colly.Response, err error) {
		log.Println("Something went wrong: ", err)
	})

	if strings.Contains(url, "https://positivnews.ru/") {
		c.OnHTML(".entry-content p", func(e *colly.HTMLElement) {
			news.text = append(news.text, e.Text)
		})

		c.OnHTML(".post-categories a", func(e *colly.HTMLElement) {
			news.category = e.Text
		})

		c.OnHTML(".entry-title", func(e *colly.HTMLElement) {
			news.title = e.Text
		})

		c.OnHTML(".entry-meta time.updated", func(e *colly.HTMLElement) {
			news.posted = formatTime(e.Attr("datetime"), "2006-01-02T15:04:05-07:00")
		})

		c.OnHTML("div.post-inner div.post-thumbnail img.wp-post-image", func(e *colly.HTMLElement) {
			news.image = e.Attr("src")
		})
	} else if strings.Contains(url, "https://ntdtv.ru/") {
		c.OnHTML("div[id=cont_post] p", func(e *colly.HTMLElement) {
			// Workaround for the articles without the heading <p> element in the cont_post div
			text := strings.TrimSpace(e.Text)
			if text != "" {
				news.text = append(news.text, text)
			}
		})

		c.OnHTML("header.entry-header h1", func(e *colly.HTMLElement) {
			news.title = e.Text
		})

		c.OnHTML("span.entry-date time", func(e *colly.HTMLElement) {
			news.posted = formatTime(e.Attr("datetime"), "2006-01-02 15:04:05")
		})

		c.OnHTML("link[itemprop=thumbnailUrl]", func(e *colly.HTMLElement) {
			news.image = e.Attr("href")
		})
		news.category = "Мировые новости"
	} else {
		log.Panic("The news source passed to GetNewsUrls func is not found!")
	}

	c.Visit(url)
	news.p1 = news.text[0] + " " + news.text[1]
	time.Sleep(500 * time.Millisecond)

	return news
}

func formatTime(src, inputLayout string) string {

	t, err := time.Parse(inputLayout, src)
	if err != nil {
		fmt.Println("Error parsing timestamp:", err)
		return ""
	}

	return t.Format("02-01-2006 15:04:05")
}

func removeDuplicatesAndRootSite(input []string, rootUrl string) []string {
	uniqueMap := make(map[string]bool)
	result := []string{}

	for _, s := range input {
		if !uniqueMap[s] && s != rootUrl && s != "" {
			uniqueMap[s] = true
			result = append(result, s)
		}
	}

	return result
}

func checkAndInsertItem(db *sql.DB, item NewsItem) error {
	if !dryRun {
		query := "SELECT COUNT(*) FROM news_items WHERE url = ?"
		var count int
		err := db.QueryRow(query, item.url).Scan(&count)
		if err != nil {
			return err
		}

		currentTime := time.Now()

		postedTime, err := time.Parse("02-01-2006 15:04:05", item.posted)
		if err != nil {
			return err
		}

		daysDiff := currentTime.Sub(postedTime).Hours() / 24

		if count > 0 {
			log.Printf("Item with URL '%s' already exists in the database.", item.url)
			return nil
		}

		if count == 0 && daysDiff <= 14 {
			textJSON, err := json.Marshal(item.text)
			if err != nil {
				return err
			}

			insertQuery := `
		INSERT INTO news_items (url, category, posted, title, image, text, p1, item_was_sent)
		VALUES (?, ?, ?, ?, ?, ?, ?, ?)
	`
			_, err = db.Exec(insertQuery, item.url, item.category, item.posted, item.title, item.image, textJSON, item.text[0], false)
			if err != nil {
				return err
			}

			log.Printf("Item with URL '%s' inserted into the database.", item.url)
		}
	} else {
		log.Printf("DRY-RUN: Item with URL '%s' inserted into the database.", item.url)
	}
	return nil
}

func processUnsentItems(db *sql.DB) error {
	if !dryRun {
		tx, err := db.Begin()

		if err != nil {
			return err
		}
		defer tx.Rollback()

		query := "SELECT id, category, posted, title, image, p1 FROM news_items WHERE item_was_sent = false"
		rows, err := tx.Query(query)
		if err != nil {
			return err
		}
		defer rows.Close()

		var unsentItems []NewsItem

		for rows.Next() {
			var id int
			var category, posted, title, image, p1 string

			if err := rows.Scan(&id, &category, &posted, &title, &image, &p1); err != nil {
				return err
			}

			item := NewsItem{
				id:       id,
				category: category,
				posted:   posted,
				title:    title,
				image:    image,
				p1:       p1,
			}

			unsentItems = append(unsentItems, item)
		}

		if len(unsentItems) == 0 {
			log.Println("No unsent items found...")
			return nil
		}

		sort.Slice(unsentItems, func(i, j int) bool {
			timeI, _ := time.Parse("02-01-2006 15:04:05", unsentItems[i].posted)
			timeJ, _ := time.Parse("02-01-2006 15:04:05", unsentItems[j].posted)
			return timeI.Before(timeJ)
		})

		for _, item := range unsentItems {
			err := sendToExternalService(item)
			if err != nil {
				log.Printf("Error sending item with ID %d to external service: %v", item.id, err)
				continue
			}

			updateQuery := "UPDATE news_items SET item_was_sent = true WHERE id = ?"
			_, err = tx.Exec(updateQuery, item.id)
			if err != nil {
				log.Printf("Error updating item_was_sent flag for item with ID %d: %v", item.id, err)
				tx.Rollback()
				continue
			}
		}

		if err := tx.Commit(); err != nil {
			return err
		}
	} else {
		log.Printf("DRY-RUN: Finished processUnsentItems")
	}

	return nil
}

func sendToExternalService(item NewsItem) error {
	apiKey := os.Getenv("API_KEY")
	chatId := os.Getenv("CHAT_ID")

	messageEnding := pickRandomMessageEnding()
	caption := url.QueryEscape(fmt.Sprintf("%s: %s\n\n%s\n\n%s\n\n%s", item.category, item.title, item.p1, item.posted, messageEnding))

	url := fmt.Sprintf("https://api.telegram.org/bot%s/sendPhoto", apiKey)

	resp, err := http.Get(fmt.Sprintf("%s?chat_id=%s&photo=%s&caption=%s&parse_mode=Markdown", url, chatId, item.image, caption))
	if err != nil {
		return err
	}

	if resp.StatusCode != http.StatusOK {
		body, err := io.ReadAll(resp.Body)
		if err != nil {
			return err
		}
		log.Printf("Request: %s?chat_id=%s&photo=%s&caption=%s&parse_mode=Markdown", url, chatId, item.image, caption)
		log.Printf("Response body:\n%s", body)
		return fmt.Errorf("sendToExternalService request for item id: %d. StatusCode: %d", item.id, resp.StatusCode)
	}
	defer resp.Body.Close()

	return nil
}

func pickRandomMessageEnding() string {
	messageEndings := []string{
		"[Подписывайся! У нас только хорошие новости!](t.me/nomoredoomscrolling)",
		"[Жми сюда, если надоел Doom Scrolling](t.me/nomoredoomscrolling)",
		"[Если понравилось, заходи. У нас есть ещё!](t.me/nomoredoomscrolling)",
		"[Ждём тебя на нашем канале!](t.me/nomoredoomscrolling)",
		"[Не упусти свою порцию позитива!](t.me/nomoredoomscrolling)",
		"[Брось Doom Scrolling и присоединяйся!](t.me/nomoredoomscrolling)",
		"[У нас всегда только светлая сторона новостей!](t.me/nomoredoomscrolling)",
		"[Забудь о плохих новостях на нашем канале!](t.me/nomoredoomscrolling)",
		"[Больше хороших новостей на нашем канале!](t.me/nomoredoomscrolling)",
		"[С нами ты всегда найдёшь причину улыбнуться!](t.me/nomoredoomscrolling)",
		"[Подними себе настроение на нашем канале!](t.me/nomoredoomscrolling)",
		"[Позитивные истории ждут тебя! Присоединяйся!](t.me/nomoredoomscrolling)",
	}

	rand.New(rand.NewSource(time.Now().UnixNano()))
	randomIndex := rand.Intn(len(messageEndings))

	return messageEndings[randomIndex]
}
